pkg/webhook/core.oam.dev/v1alpha2/applicationconfiguration/validating_handler.go

```diff
@@ -210,6 +215,74 @@ func ValidateTraitAppliableToWorkloadFn(_ context.Context, v ValidatingAppConfig
 	return allErrs
 }
 
+// ValidateTraitConflictFn validates whether conflicting traits are applied to the same workload.
+// NOTE(roywang) It returns immediately if one conflict is detected
+// instead of returning after collecting ALL conflicts
+func ValidateTraitConflictFn(_ context.Context, v ValidatingAppConfig) []error {
+	klog.Info("validate trait conflicts ", "appconfig name:", v.appConfig.Name)
+	allErrs := make([]error, 0)
+	for _, comp := range v.validatingComps {
+		allConflictRules := map[string][]string{}
+		// collect conflicts rules of all traits applied to this workload
+		for _, trait := range comp.validatingTraits {
+			allConflictRules[trait.traitDefinition.Name] = trait.traitDefinition.Spec.ConflictsWith
+		}
+
+		for rulesOwner, rules := range allConflictRules {
+			if len(rules) == 0 {
+				// empty rules means this trait can work with any other ones
+				continue
+			}
+			noConflictWithAll := false
+			for _, rule := range rules {
+				if rule == "*" {
```

Comment:
if we write a `*` as conflict rule, this might mean the trait is conflict to all other traits, right?

Reply:
Yes, that makes sense. I originally treated "*" as an explicit declaration of empty rule set. Let me remove it.